### PR TITLE
Make compaction size-bounded and share one engine

### PR DESCRIPTION
**Pull request by tjgreen42**
## Summary

Make segment compaction size-bounded, and share one planner and executor
between ordinary threshold compaction and `bm25_force_merge()`.

Compaction now combines the largest adjacent groups of segments that fit
`pg_textsearch.max_segment_size` (new GUC, default 4095MB) instead of merging
a whole level unconditionally. A segment that already exceeds the budget is
left alone as an indivisible singleton.

This establishes the storage behavior required by later background compaction
work, but contains no scheduler, transaction-dispatch, or pg_durable
integration.

## Why

BM25 indexes store immutable segments across levels L0 through L7. `L7` is
terminal.

Force merge was documented as analogous to Lucene's `forceMerge(1)`: rewrite
all segments into a single segment. That postcondition is not reachable in
general. A corpus larger than one representable segment can never satisfy it,
so on a large index force merge either failed or could not do the useful
consolidation it was still capable of. Merging an entire level in one batch
has the same problem: batch size is unbounded and driven by whatever happens
to be on the level.

Bounding the *size* of a merge output rather than the *count* of resulting
segments removes the unreachable postcondition. Once compaction is bounded,
force merge is just ordinary compaction with a different trigger, so the two
previously separate paths in `merge.c` collapse into one engine.

## What changed

- Add `src/segment/compaction.c`, a shared bounded planner and executor used
  by both ordinary threshold compaction and force merge.
- Add `pg_textsearch.max_segment_size` (default 4095MB, range 1-4095MB): the
  conservative size budget for newly merged multi-source segments.
- Plan merges as the largest adjacent groups that fit the budget, rather than
  one batch per level.
- Leave an over-budget segment as an indivisible singleton instead of failing
  the operation.
- Keep searching higher levels when a triggered level has no legal count
  reduction, so an indivisible L0 cannot starve a compactable L1.
- Centralize pre-mutation checks for persisted per-level segment counts, so a
  level that cannot accept another segment fails before any page is mutated.
  Page changes are WAL-logged with `GenericXLog` and do not roll back with the
  surrounding transaction, so this check has to happen up front.
- Require callers to hold the per-index exclusive lock;
  `tp_require_compaction_lock()` asserts it and never acquires or releases, so
  a lock inherited from spill or build scope is never dropped early.
- Reject `bm25_force_merge()` during recovery, matching
  `bm25_spill_index()`. It had no such guard, so on a standby it returned
  success whenever the plan was a no-op, reporting compaction that had not
  happened.
- Add a testing-only segment-count limit GUC so capacity boundaries can be
  exercised without constructing 65,535 segments.

## User-visible behavior

Ranking and query results are unchanged.

`bm25_force_merge()` is no longer `forceMerge(1)`. It performs one bounded
compaction pass and **does not guarantee a single segment**: large indexes may
intentionally retain several immutable segments, and that is a success rather
than a partial result. Segments that individually exceed the budget remain
indivisible. README and CLAUDE.md are updated to state this contract.

Existing behavior is otherwise preserved: sources stay readable and immutable
until the replacement is published, and pages displaced by publication
continue to enter the deferred-free tombstone chain (#380) rather than
returning to the FSM immediately.

This addresses the force-merge half of
[#456](https://github.com/timescale/pg_textsearch/issues/456) (bounded merges
for corpora that cannot fit one segment).
[#473](https://github.com/timescale/pg_textsearch/issues/473) still tracks
consistent fail-safe enforcement of the per-segment format limits.

## Test coverage

- Bounded ordinary compaction produces maximal batches without exceeding the
  configured budget.
- Bounded force merge succeeds while retaining multiple segments, including
  over-budget indivisible singletons.
- A level with no legal count reduction does not starve a higher level that
  has one.
- Full destination levels are drained before lower levels promote.
- Segment-count exhaustion rejects without changing the index.
- Empty, all-dead, termless-memtable, and ordinary live-document cases
  preserve document accounting and query results.
- Concurrency: `vacuum_concurrent_merge.sh` no longer tolerates force-merge
  capacity errors, and `partial_concurrent_read.sh` drives two concurrent
  force-merge workers against one index.
- Replication: a standby rejects both `bm25_spill_index()` and
  `bm25_force_merge()` with a read-only error.

## Stack position

This is the bottom, independently mergeable layer of GitHub stack #479.
Later layers add the public APIs, transaction dispatch, and optional
pg_durable integration.

### CODE REVIEW

File: test/scripts/vacuum_concurrent_merge.sh [hunk 1]
```diff
@@ -137,10 +137,22 @@ writer() {
...
-        $PSQL -c "SELECT bm25_spill_index('docs_bm25');
-                  SELECT bm25_force_merge('docs_bm25')" \
-          >>"${ERR_DIR}/merger.log" 2>&1 || return 30
+        if ! output=$($PSQL -c \
+            "SELECT bm25_spill_index('docs_bm25');
+             SELECT bm25_force_merge('docs_bm25')" 2>&1); then
+            printf '%s\n' "$output" >>"${ERR_DIR}/merger.log"
+            if [[ "$output" == \
+*"level 7 is occupied while lower levels remain"* ]] \
+                || [[ "$output" == \
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply test/scripts/vacuum_concurrent_merge.sh]: Fixed in 9798d4e9261ca86cacf45e1bd69525bc6ce1f8f0. The test now matches the current `level 7 is occupied while spilling the memtable would create another segment` guard exactly. I kept exact message matching rather than broadening to SQLSTATE 54000: `tp_force_merge_preflight` has several other program-limit failures, and default `psql` output does not include the SQLSTATE, so a broader match would either be ineffective or risk hiding unrelated errors. Validated with `bash -n`, the force-merge source invariant, PostgreSQL 17.10 `installcheck` (75/75), and full-scale `vacuum_concurrent_merge.sh`.
